students/Kate_Koehn/Session06/mailroom_4.py

```diff
@@ -0,0 +1,172 @@
+def get_sample_donors():
+    donors = {"Rick Sanchez": [3.00, 1.00],
+              "Liz Lemon": [4000.00, 3000.00, 6000.00],
+              "Andy Dwyer": [10.00],
+              "Brendan Small": [3.00, 10.00],
+              "Coach McGuirk": [2.00, 1.00]}
+    return donors
+
+# name of folder in which to store thank you letters to donors
+OUT_PATH = "thank_you_letters"
+
+
+def prepare_to_run():
+    """checks if folder in which to store thank you letter files exists, if not, creates new folder"""
+    if not os.path.isdir(OUT_PATH):
+        os.mkdir(OUT_PATH)
+
+
+def quit():
+    """quits the program"""
+    sys.exit(0)
+
+
+def menu_selection():
+    '''main function to run the menus from'''
+    user_selection = input(dedent('''
+          Hello! Please tell us what you'd like to do:
+          1 - Send a Thank You
+          2 - Create a Report
+          3 - Send Letters to All Donors
+          4 - Quit
+          > '''))
+    return user_selection.strip()
+
+
+def create_report():
+    """creates a report of all previous donors and their donations"""
+    report = []
+    report.append("{:<20}{:5}{:5}{}".format("Donor Name", "| Total Given", "| Num Gifts", "| Average Gift"))
+    report.append("{:-<60}".format(""))
+    for donor_name, gifts in donors.items():
+        total = sum(gifts)
+        num_gifts = len(gifts)
+        avg_gift = total / num_gifts
+        report.append("{:<22} ${:<13.2f} {:<10} ${:.2f}".format(donor_name, total, num_gifts, avg_gift))
+    return "\n".join(report)
+
+
+def email(donor_name, donation_amount):
+    """creates individual thank-you email for a new donation"""
+    message = ("\n\nDear {},"
+               "\n\nThanks for your money! Your donation of ${:.2f} will be summarily used to buy me beer.\n\n"
+               "Kind regards,\nKate Koehn").format(donor_name, donation_amount)
+    file_name = "{}.txt".format(donor_name)
+    with open(file_name, "w") as f:
+        f.write(message)
+
+
+def generate_letter(donor_name, donation_amount):
+    """generates a thank you letter for the donation made"""
+    return dedent('''Dear {!s:s},
+
+    Thank you for your crappy donation of ${:.2f}. It will be used to buy the cheapest beer possible, because you couldn't donate enough for us to buy anything decent. We'll be lucky if we get Natty Ice.
+
+    Cheers,
+        Kate Koehn
+    '''.format(donor_name, donation_amount))
+
+
+def get_donor_key(donor_name):
+    """looks up key in donor dict"""
+    donor_key = donor_name.strip()
+    for key in donors.keys():
+        if donor_key == key:
+            return True
+    return False
+
+
+def add_new_donor(donor_name):
+    name = donor_name.strip()
+    if name not in donors.keys():
+        donors[name] = []
+    return donor_name
+
+
+def add_donation_to_donor(donor_name, donation):
+    donors[donor_name].append(donation)
+
+
+def check_donation(donation):
+    donation = float(donation)
```

Comment:
this could use a docstring (they all could :-) )

"""
checks if a string for a donation is a valid float.

:param donation: the donation string

:returns: a float with the value.

Raises a ValueError if donation is not a valid value
"""